### PR TITLE
Add basic device APIs to the top-level `torch_xla` module.

### PR DESCRIPTION
**Pull request by will-cromar**
Implemented basic device APIs from #6399.

- Add `device`, `devices`, `real_devices`, and `device_count` as described in RFC.
- We already do a substantial amount of setup in `__init__.py` . Add `torch_xla.py` for public functions on `torch_xla` module.

Follow up:
- Update documentation to use new APIs.
- Start deprecating or discouraging usage of old APIs like `xm.xla_device` and `runtime.xla_device`.

### CODE REVIEW

File: torch_xla/csrc/init_python_bindings.cpp [hunk 1]
```diff
@@ -1188,14 +1188,20 @@ void InitXlaModuleBindings(py::module m) {
...
-  });
+  m.def("_xla_real_devices",
+        [](const std::optional<std::vector<std::string>> devices) {
+          if (!devices) {
+            return runtime::GetComputationClient()->GetLocalDevices();
+          }
+
+          std::vector<std::string> xla_devices;
+          {
+            NoGilSection nogil;
```
Comment (vanbasten23) [inline torch_xla/csrc/init_python_bindings.cpp:1199]: Do you know why we release the GIL in this block? Is it for tpu v2/v3 where we allow multiple threads to do some runtime job such as `GetXlaDevices(*devices)`?

Comment (will-cromar) [reply torch_xla/csrc/init_python_bindings.cpp:1199]: I'm not sure to be honest. Maybe `GetLocalDevices` was a blocking call in XRT?

File: test/test_devices.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import torch
+import torch_xla as xla
+import torch_xla.runtime as xr
+
+
+class TestDevices(parameterized.TestCase):
+
+  def setUpClass():
+    xr.set_device_type('CPU')
+    os.environ['CPU_NUM_DEVICES'] = '4'
```
Comment (JackCaoG) [inline test/test_devices.py:13]: hmm, it is OK for now but shouldn't we also test it on GPU and TPU?

File: torch_xla/__init__.py [hunk 3]
```diff
@@ -182,3 +182,5 @@ def _init_xla_lazy_backend():
 if os.getenv('XLA_REGISTER_INSTALLED_PLUGINS') == '1':
   plugins.use_dynamic_plugins()
   plugins.register_installed_plugins()
+
+from .torch_xla import *
```
Comment (JackCaoG) [inline torch_xla/__init__.py:186]: why is this needed?

File: test/test_devices.py (hunk 2 above)
Comment (will-cromar) [reply test/test_devices.py:13]: This is sufficient IMO. We're really just testing the integration of this module with the runtime client, which has the same API regardless of the underlying device.

As we switch to using these functions by convention, they'll be exercised by almost every other test.

File: torch_xla/__init__.py (hunk 3 above)
Comment (will-cromar) [reply torch_xla/__init__.py:186]: This imports the contents of `torch_xla.py` into `torch_xla/`'s module scope. Otherwise, the functions would be `torch_xla.torch_xla.etc`. This assigns them to `torch_xla.etc`

File: torch_xla/csrc/init_python_bindings.cpp (hunk 1 above)
Comment (JackCaoG) [reply torch_xla/csrc/init_python_bindings.cpp:1199]: oh lol it is that `torch_xla`

**Review by JackCaoG — APPROVED** [review]
